presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1159,8 +1165,15 @@ protected Scope visitSetOperation(SetOperation node, Optional<Scope> scope)
             Type[] outputFieldTypes = relationScopes.get(0).getRelationType().getVisibleFields().stream()
                     .map(Field::getType)
                     .toArray(Type[]::new);
+            int outputFieldSize = outputFieldTypes.length;
+            if (isExpensiveUnionDistinct(node, outputFieldTypes)) {
+                warningCollector.add(new PrestoWarning(
+                        PERFORMANCE_WARNING,
+                        format("UNION DISTINCT query should consider reducing the number of visible fields (%d) to %d or less",
```

Comment:
My apologies. I should've been more thorough before. The error message didn't mention types. So it could be confusing if the warning is actually triggered by using DOUBLE / complex types rather than exceeding 3 fields. So let's mention both here.